README.md
# Domain Renewal and DynDNS for Freenom.com

[![Shellcheck](https://github.com/mkorthof/freenom-script/actions/workflows/shellcheck.yml/badge.svg)](https://github.com/mkorthof/freenom-script/actions/workflows/shellcheck.yml)
[![Docker](https://github.com/mkorthof/freenom-script/actions/workflows/docker.yml/badge.svg)](https://github.com/mkorthof/freenom-script/actions/workflows/docker.yml)
[![BATS](https://github.com/mkorthof/freenom-script/actions/workflows/bats.yml/badge.svg)](https://github.com/mkorthof/freenom-script/actions/workflows/bats.yml)

## Last Update ##

⚠ Freenom recently added CAPTCHA from AWS WAF, causing this script to stop working (June 2023).

***Latest version: v2023-08-03 ([CHANGES.md](CHANGES.md))***

**Make sure to add new config options when updating script**

---

This shell script makes sure your Freenom domains don't expire by auto renewing them. It can also set a DNS record with the clients ip address.

You'll need to have already registered an account at Freenom.com with at least one (free) domain added, before you can run the script.

## Usage

```shell

freenom.com Domain Renewal and DynDNS
-------------------------------------

Usage:      freenom.sh -l [-d]
            freenom.sh -r <domain OR -a> [-s <subdomain>]
            freenom.sh -u <domain> [-s <subdomain>] [-m <ip>] [-f]
            freenom.sh -z <domain>

Options:    -l    List all domains and id's in account
                  add [-d] to show renewal Details
            -r    Renew <domain> or use '-r -a' to update All
                  add [-s] to renew <Subdomain>
            -u    Update <domain> A record with current ip
                  add [-s] to update <Subdomain> record
                  add [-m <ip>] to Manually update static <ip>
                  add [-f] to Force update on unchanged ip
            -z    Zone for <domain>, shows dns records

            -4    Use ipv4 and modify A record on "-u" (default)
            -6    Use ipv6 and modify AAAA record on "-u"
            -c    Config <file> to use, instead of freenom.conf
            -i    Ip commands list, used to get current ip
            -o    Output renewals, shows html file(s)

Examples    ./freenom.sh -r example.com
            ./freenom.sh -c /etc/mycustom.conf -r -a
            ./freenom.sh -u example.com -s mail

            * When "-u" or "-r" is used with argument <domain>
              any settings in script or config file are overridden

```

## Installation

Using a full Linux distro including coreutils is recommended (e.g. Debian). Embedded and BusyBox based systems are untested and will probably not work correctly or at all.

_Note that this shell script requires recent versions of "Bash" and "cURL"_

### Auto Installer

Run: `make install` 

(from git clone directory)

This automatically installs the script, .conf file and configures scheduler.

### Manual install

Suggested installation path: "/usr/local/bin/freenom.sh"

And for the config file: "/usr/local/etc/freenom.conf"

### Docker

There's an [image](https://github.com/mkorthof/freenom-script/pkgs/container/freenom_script) available from GitHub Container Registry which you can run like this:

 `docker run --rm --env freenom_email="you@example.com" --env freenom_passwd="yourpassword" ghcr.io/mkorthof/freenom_script -l`

For more information see [Docker.md](docs/Docker.md)

## Configuration

Settings can be changed in the script itself or set in a separate config file (default). Every setting has comments with possible options and examples.

First edit config and **set your email and password** which you use to sign-in to freenom.com

- The default filename is "freenom.conf" in the same location as the script, or "/usr/local/etc/freenom.conf"
- You can also use `freenom.sh -c /path/to/file.conf`
- To optionally put config in the script itself instead: copy settings from conf file into `freenom.sh` (before "Main")

Default settings such as retries and timeouts can be changed, they are however OK to leave as-is (see [Overrides.md](docs/Overrides.md)).

### Testing

Test the script by running `freenom.sh -l` and make sure your domains are listed. To Renew domains or Update A record, see [Usage](#usage) or `freenom.sh -h`  

## Scheduling

Optionally you can schedule the script to run automatically. The installer creates "/etc/cron.d/freenom" or systemd timers in 'system mode' so the script runs at certain intervals. It will output a message with instructions on how to set your domain(s) to renew/update or renew all:

- Cron:
    - edit the created file in /etc/cron.d and uncomment line(s)
- Systemd:
    - `systemctl enable --now freenom-renew-all.timer`
    - `systemctl enable --now freenom-update@example.tk.timer`
    - `systemctl enable --now freenom-update@mysubdom.example.tk.timer`
    
_If systemd is not available on your system, the installer will use cron instead._

### Manual setup

See [Scheduling.md](docs/Scheduling.md)

## Notifications

### Email

To enable email alerts, make sure `MTA` is set; the default is `/usr/sbin/sendmail`. Emails will be sent on renewal or update errors. If you do not have/want an MTA installed you could use [bashmail](https://git.io/JJdto) instead.

If you want to receive the alerts on a different email address than `freenom_email` set `RCPTTO`. You can also set an optional "From" address: `MAILFROM="Freenom Script <freenom-script@example.tk>"`

Leaving `MTA` empty or commented disables alerts.

### Apprise

Uses external lib to send notification to many services like Telegram, Discord, Slack, Amazon SNS, MS Teams etc.

To enable [Apprise](https://github.com/caronc/apprise) notifications, make sure `APPRISE` is set to the location where you installed the Apprise CLI; the default is `/usr/local/bin/apprise`. You must also set the `APPRISE_SERVER_URLS` array to contain one or more server URLs. Notifications are sent to all of the listed server URLs. As with email notifications, Apprise notifications are sent on renewal or update errors.

For details on how to construct server URLs, refer to [supported notifications](https://github.com/caronc/apprise#supported-notifications).

Leaving the `APPRISE_SERVER_URLS` array empty disables Apprise notifications. 

## DynDNS

To update A or AAAA records the nameservers for the domain must be set to the default Freenom Name Servers.

The record will be added if there is none, or modified if the record already exists. Your current ip address will be used as value (aka "Target").

### IP Address

The script uses 3 methods to get your current ip address from a number of public services:

- HTTP method: `curl https://checkip.amazonaws.com`
- DNS method: `dig TXT +short o-o.myaddr.l.google.com @ns1.google.com`
- Manually: you can also set a static ip address instead of auto detect (see below)

There are a few more HTTP and DNS services defined for redundancy, the script will choose one at random. By default it will retry to get the ip 3 times.

Once your ip is found it's written to "freenom_\<domain\>.ip4" (or 'ip6'). Same if freenom returns dnserror "There were no changes". This is to prevent unnecessary updates in case the ip is unchanged.
To force an update you can remove this file which is located in the default output path: "/var/log".

To manually update: set `freenom_static_ip=<your ip>` and `freenom_update_manual="1"`, or use the `-m` option.

### Issues

#### Timeout

In case of dns issues where resolving does not work correctly and curl times out to https://my.freecom.com ('exit code 28'), try setting `freenom_http_resolve="1"`. This make sure curl uses Freenoms own public dns servers.

#### Account

You need an actual Freenom account, as using Social Sign-in will not work. Workaround: use password reset, see [KB](https://my.freenom.com/knowledgebase.php?action=displayarticle&id=27) and issue [#56](https://github.com/mkorthof/freenom-script/issues/56).

#### Get ip

Make sure 'curl' and/or 'dig' is installed (e.g. debian: dnsutils or redhat: bind-utils). In case of issues try running curl and dig command manually.

- To list all 'get ip' commands run `freenom.sh -i` (or `grep getIp freenom.conf`)
- To disable IPv6: set `freenom_update_ipv="4"`
- To disable dig: set `freenom_update_dig="0"`

## Files

- Installer: Makefile (`make install`)
- Script/config: `freenom.sh` and `freenom.conf`
- Output:
  - Path: `"/var/log/freenom/"` (default)
  - Files: `freenom.log`, `freenom_<domain>.ip{4,6}`, `freenom_renewalResult-<id>.html`
- View Results: use `freenom.sh -o` to view html files
  
Also see comment "Output files" and `freenom_out_dir` variable in conf.

### Updating

Usually you can just replace "freenom.sh" with the new version, if you're using a separate config file.

An exception is when config options were added/changed which you may need to compare and merge. Such config changes are listed in [CHANGES.md](CHANGES.md).

### Uninstall

Run `make uninstall`.

You can also manually reverse the steps under [Installation](#installation) above (e.g. remove .sh, .conf and scheduler files).

## Sources

See included [orig](orig) dir

- Original script: [gist.github.com/a-c-t-i-n-i-u-m/bc4b1ff265b277dbf195](https://gist.github.com/a-c-t-i-n-i-u-m/bc4b1ff265b277dbf195)
- Updated script: [gist.github.com/pgaulon/3a844a626458f56903d88c5bb1463cc6](https://gist.github.com/pgaulon/3a844a626458f56903d88c5bb1463cc6)
- Reference: [github.com/dabendan2/freenom-dns](https://github.com/dabendan2/freenom-dns) (nodejs/npm)
- Reference: [github.com/patrikx3/freenom](https://github.com/patrikx3/freenom) (nodejs/npm)

## Changes

See [CHANGES.md](CHANGES.md)


CHANGES.md
# Changes

_order: latest/newest on top_

- [**20230803**] fix aws token validation  ([#68](https://github.com/mkorthof/freenom-script/issues/68))
- [**20230610**] added env var AWS_WAF_TOKEN ([#66](https://github.com/mkorthof/freenom-script/issues/66))
- [**20221224**] added option to make curl use public freenom dns
  - **config changes:**
    - `curlDns="80.80.80.80,80.80.81.81"`
    - `resolveCmd+=("dig +short @80.80.80.80 my.freenom.com")`
    - `resolveCmd+=("nslookup my.freenom.com 80.80.80.80")`
    - `resolveCmd+=("host my.freenom.com 80.80.80.80")`
    - renamed var `c_exOpts` to `curlExtraOpts`
- [**20221217**] fixed grep3.8 warnings (PR [#63](https://github.com/mkorthof/freenom-script/pull/63) from aleksav013)
- [**20220812**] added bash/busybox check, log login errors
- [**20220806**] add docker image ([#57](https://github.com/mkorthof/freenom-script/issues/57))
  - **config change:** `uaString`
- [**20220626**] update ip file on freenom dnserror "There were no changes"
- [**20220325**] added pid to logs, **config changes:** ipCmd uaString
- [**20220212**] changed systemd unit file to support updating subdomains (PR [#54](https://github.com/mkorthof/freenom-script/pull/54))
- [**20220108**] added support for [Apprise](https://github.com/caronc/apprise) (PR [#53](https://github.com/mkorthof/freenom-script/pull/53) from jjlin)
  - **config changes:**
    - `APPRISE="/usr/local/bin/apprise"` 
    - `APPRISE_SERVER_URLS=( )`
- [**20211207**] fixed issue [#52](https://github.com/mkorthof/freenom-script/issues/52): sed removes CR using \r instead of ^M
- [**20210821**] fixed issue [#50](https://github.com/mkorthof/freenom-script/issues/50)
- [**20210618**] added more config checks (PR [#43](https://github.com/mkorthof/freenom-script/issues/43) from nathanfranke)
- [**20210612**] check for uaString and ipCmd settings ([#42](https://github.com/mkorthof/freenom-script/issues/42))
- [**20210508**] fixed issue with `-c` option ([#41](https://github.com/mkorthof/freenom-script/issues/41))
- [**20210507**] fixed renew all systemd unit (PR [#39](https://github.com/mkorthof/freenom-script/issues/39) from aeolyus)
- [**20200927**] fixed mistake unsetting a var ([#34](https://github.com/mkorthof/freenom-script/issues/34))
- [**20200918**] fixed installer conf path issue [#33](https://github.com/mkorthof/freenom-script/issues/33)
- [**20200815**] fixed ip6 issue [#32](https://github.com/mkorthof/freenom-script/issues/32) and an ipv6check, added email alerts
  - **config changes:**
    - `MTA="/usr/sbin/sendmail"`
    - `RCPTTO="admin@example.tk"`
    - `MAILFROM="Freenom Script <freenom-script@example.com>"`
- [**20200713**] fixed systemd in installer and mistakes in readme (and [#31](https://github.com/mkorthof/freenom-script/issues/31))
- [**20200705**] mainly fixes:
  - fixed ip update on apex record
  - fixed conf path detection
  - fixed (older) curl failing login ([#25](https://github.com/mkorthof/freenom-script/issues/25))
  - fixed missing expirydate ([#28](https://github.com/mkorthof/freenom-script/issues/28))
  - fixed dot in subdomain aka "sub subdomain" ([#29](https://github.com/mkorthof/freenom-script/issues/29))
  - added (test) option: update all domains/records '-a' ([#19](https://github.com/mkorthof/freenom-script/issues/19))
  - improved logging update errors
  - **config changes:**
    - `freenom_http_sleep="3 3"`
    - `freenom_oldcurl_force="0"`
- [**20200312**] added checks for required bins
- [**20200307**] fixed Makefile, installing freenom.sh was disabled
- [**20200128**] fixed escaping special chars in password ([#21](https://github.com/mkorthof/freenom-script/issues/21))
- [**20200127**] fixed domain renewals ([#23](https://github.com/mkorthof/freenom-script/issues/23))
- [**20191127**] always cleanup cookie file
- [**20191125**] update all domains ([#19](https://github.com/mkorthof/freenom-script/issues/19))
  - **config change:** `freenom_update_all="0"`
- [**20191108**] static ip update ([#18](https://github.com/mkorthof/freenom-script/issues/18))
  - **config changes:**
    - `freenom_static_ip=""`
    - `freenom_update_manual="0"`
- [**20191017**] changed default conf path to /usr/local/etc
- [**20191005**] errorUpdateResult is no longer saved to html file
- [**20190931**] added systemd templates (PR [#15](https://github.com/mkorthof/freenom-script/issues/15) from sdcloudt)
- [**20190927**] added installer
- [**20190922**] fixed issue [#12](https://github.com/mkorthof/freenom-script/issues/12) (sdcloudt)
- [**20190920**] changed out dir
  - **config change:** `freenom_out_dir="/var/log/freenom"`
- [**20190920**] added curl retries
  - **config change:** `freenom_http_retry="3"`
- [**20190621**] changed 'cannot renew until' warnings to notices
- [**20190621**] added tests (BATS)
- [**20190621**] added recType check (A or AAAA) for dyndns
- [**20190420**] option to use separate .conf file
- [**20190420**] option to skip renewal notice details in log
- [**20190420**] option to list existing dns records
- [**20190317**] retry getting current ip
- [**20190317**] handling of existing dns records (ip update)
- [**20190000**] made updating ip optional
- [**20190000**] fixed token
- [**20190000**] added referer url to curl to fix login
- [**20190000**] added logging
- [**20190000**] added domain renewals

More details: `git log --pretty=short --name-only`


test/test.sh
#!/bin/sh

# test 1 function:
# bats freenom-funcs.bats -f func_getRec

func_msg () {
  test -n "$1" && echo "$(date '+%F %H:%M:%S') ${1}: test.sh"
}

case $1 in
  args)   trap 'func_msg "END"; exit' EXIT HUP INT TERM
          func_msg "START"
          bats freenom-args.bats
          ;;
  funcs)  trap 'func_msg "END"; exit' EXIT HUP INT TERM
          func_msg "START"
          bats freenom-funcs.bats
          ;;
  sc)     trap 'func_msg "END"; exit' EXIT HUP INT TERM
          func_msg "START"
          bats freenom-sc.bats
          ;;
  all)    func_msg "START"
          trap 'func_msg "END"; exit' EXIT HUP INT TERM
          if [ -d ./test ]; then
            ( bats -r ./test/freenom* -t )
          elif [ -f "freenom-args.bats" ] && [ -f "freenom-funcs.bats" ] && [ -f "freenom-sc.bats" ]; then
            ( cd .. && bats -r ./test/freenom* -t )
          else
            echo "ERROR: bats files not found"
            exit 1
          fi
          ;;
  *)      echo "$(basename "$0") <args|funcs|sc|all>"
          exit 0
          ;;
esac


orig/freenom.com.ddns.sh
#!/bin/bash
 
# settings
# Login information of freenom.com
freenom_email="main@address"
freenom_passwd="pswd"
# Open DNS management page in your browser.
# URL vs settings:
#   https://my.freenom.com/clientarea.php?managedns={freenom_domain_name}&domainid={freenom_domain_id}
freenom_domain_name="domain.name"
freenom_domain_id="000000000"
 
# main
# get current ip address
current_ip="$(curl -s "https://api.ipify.org/")"
 
if [ "${current_ip}" == "" ]; then
    echo "Could not get current IP address." 1>&2
    exit 1
fi

# login
cookie_file=$(mktemp)
loginResult=$(curl --compressed -k -L -c "${cookie_file}" \
    -F "username=${freenom_email}" -F "password=${freenom_passwd}" \
    "https://my.freenom.com/dologin.php" 2>&1)

if [ "$(echo -e "${loginResult}" | grep "/clientarea.php?incorrect=true")" != "" ]; then
    echo "Login failed." 1>&2
    exit 1
fi

exit 0

# update
updateResult=$(curl --compressed -k -L -b "${cookie_file}" \
    -F "dnsaction=modify" -F "records[0][line]=" -F "records[0][type]=A" -F "records[0][name]=" -F "records[0][ttl]=14440" -F "records[0][value]=${current_ip}" \
    "hxxttps://my.freenom.com/clientarea.php?managedns=${freenom_domain_name}&domainid=${freenom_domain_id}" 2>&1)

if [ "$(echo -e "$updateResult" | grep "<li class=\"dnssuccess\">")" == "" ]; then
    echo "Update failed." 1>&2
    exit 1
fi

# logout
curl --compressed -k -b "${cookie_file}" "https://my.freenom.com/logout.php" > /dev/null 2>&1

# clean up
rm -f ${cookie_file}

exit 0


orig/freenom.com.ddns-updated.sh
#!/bin/bash

#
# settings
#
# login data
freenom_email="main@address"
freenom_passwd="pswd"
# Open DNS management page in your browser.
# URL vs settings:
#   https://my.freenom.com/clientarea.php?managedns={freenom_domain_name}&domainid={freenom_domain_id}
freenom_domain_name="domain.name"
freenom_subdomain_name=""
freenom_domain_id="000000000"

#
# main
#
# get current url
current_ip="$(dig TXT +short o-o.myaddr.l.google.com @ns1.google.com 2>/dev/null | tr -d '"')"
log_file="/tmp/$(basename $0)"

if [ "$current_ip" == "" ]; then
    echo "[$(date)] Couldn't get current global ip address." >> ${log_file}.log
    exit 1
fi
if [ "$(cat ${log_file}.ip 2>/dev/null)" == "$current_ip" ]; then
    exit 0
fi

# login
cookie_file=$(mktemp)

loginPage=$(curl --compressed -k -L -c "$cookie_file" \                                                                                                                            
    "https://my.freenom.com/clientarea.php" 2>&1)

token=$(echo "$loginPage" | grep token | grep -o value=".*" | sed 's/value=//g' | sed 's/"//g' | awk '{print $1}')

loginResult=$(curl --compressed -k -L -c "$cookie_file" \
    -F "username=$freenom_email" -F "password=$freenom_passwd" -F "token=$token"\
    "https://my.freenom.com/dologin.php" 2>&1)

if [ "$(echo -e "$loginResult" | grep "Location: /clientarea.php?incorrect=true")" != "" ]; then
    echo "[$(date)] Login failed." >> ${log_file}.log
    rm -f $cookie_file
    exit 1
fi

# if record does not exists, add new record, else update the first record; records[0]
dnsManagementURL="https://my.freenom.com/clientarea.php?managedns=$freenom_domain_name&domainid=$freenom_domain_id"
dnsManagementPage=$(curl --compressed -k -L -b "$cookie_file" "$dnsManagementURL")
if [ "$(echo "$dnsManagementPage" | grep -F "records[0]")" == "" ]; then
    recordKey="addrecord[0]"
    dnsAction="add"
else
    recordKey="records[0]"
    dnsAction="modify"
fi

token=$(echo "$dnsManagementPage" | grep token | grep -o value=".*" | tail -n 1 | sed 's/value=//g' | sed 's/"//g' | awk '{print $1}') 

# request add/update DNS record
updateResult=$(curl --compressed -k -L -b "$cookie_file" \
    -F "dnsaction=$dnsAction" \
    -F "$recordKey[line]=" \
    -F "$recordKey[type]=A" \
    -F "$recordKey[name]=$freenom_subdomain_name" \
    -F "$recordKey[ttl]=14440" \
    -F "$recordKey[value]=$current_ip" \
    -F "token=$token" \
    "$dnsManagementURL" 2>&1)

# logout
curl --compressed -k -b "$cookie_file" "https://my.freenom.com/logout.php" > /dev/null 2>&1

# clean up
rm -f $cookie_file

if [ "$(echo -e "$updateResult" | grep "$current_ip")" == "" ]; then
    echo "[$(date)] Update failed." >> ${log_file}.log
    echo -e "$updateResult" > ${log_file}.errorUpdateResult.log
    exit 1
else
    # save ip address
    echo -n "$current_ip" > ${log_file}.ip
    exit 0
fi
